clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java

```diff
@@ -264,40 +271,133 @@ public boolean rename(Path src, Path dst) throws IOException {
             return false;
         }
 
-        ObjectStats srcStat;
-        ObjectsApi objects = lfsClient.getObjects();
+        LakeFSFileStatus srcStatus;
         try {
-            // Stat src file to get its metadata
-            srcStat = objects.statObject(srcObjectLoc.getRepository(), srcObjectLoc.getRef(),
-                    srcObjectLoc.getPath());
-        } catch (ApiException e) {
-            LOG.error("rename: could not get src object stats. src:{}", src, e);
+            srcStatus = getFileStatus(src);
+        } catch (FileNotFoundException e) {
+            LOG.error("rename: src {} does not exist, rename failed.", src, e);
             return false;
         }
+        if (!srcStatus.isDirectory()) {
+            return renameFile(srcStatus, dst);
+        }
+        return renameDirectory(src, dst);
+    }
 
-        return renameObject(srcStat, srcObjectLoc, dstObjectLoc);
+    /**
+     * Recursively rename objects under src dir.
+     * @return true if all objects under src renamed successfully, false otherwise.
+     */
+    private boolean renameDirectory(Path src, Path dst) throws IOException {
+        boolean dstExists = false;
+        LakeFSFileStatus dstFileStatus;
+        try {
+            // May be unnecessary with https://github.com/treeverse/lakeFS/issues/1691
+            dstFileStatus = getFileStatus(dst);
+            dstExists = true;
+            if (!dstFileStatus.isDirectory()) {
+                LOG.error("renameDirectory: dst {} is not a directory, rename of directory to a file " +
+                        "is unsupported", dst);
```

Comment:
Do we need to repeat the function name?  Doesn't the logger know it?  In any case, if we have to repeat the function name here then either don't say "directory" in the message or lie about the function name and call it "`rename`" (and I can lie to myself that that's the name of the _operation_).

```suggestion
                LOG.error("renameDirectory: dst {} is not a directory", dst);
```
or
```suggestion
                LOG.error("renameDirectory: destination {} is not a directory but source {} is", dst, src);
```
or even copy the POSIX error message from `mv`
```suggestion
                LOG.error("renameDirectory: cannot overwrite non-directory '{}' with directory '{}'", dst, src);
```

Reply:
The logger does not add the method name unless we configure it to do so. I checked this and it seems that it impacts performance so I don't think we should do it. I will change the wording per your suggestions :)